### PR TITLE
Solves a Bug in Derivative

### PR DESCRIPTION
**Pull request by Saanidhyavats**
* core
  * Replaced Derivative with cls which was related to issue #16350

### CODE REVIEW

**Comment by Saanidhyavats** [comment]
Please let me know if the changes made are wrong somewhere.

**Comment by czgdp1807** [comment]
Please DO NOT clear the PR template. @sympy-bot will give errors for RELEASE NOTES. Please correct the OP.

**Comment by Saanidhyavats** [comment]
@czgdp1807 I am unable to edit the description, shall I create a new pull request?

**Comment by czgdp1807** [comment]
I think GitHub allows the author to edit their description.

**Comment by Saanidhyavats** [comment]
I have to change the description according to sympy-bot. I am unable to find the option for changing the description. Sorry to bother @czgdp1807, can you tell me what is OP?

**Comment by jksuom** [comment]
There are three dots in the upper right corner of the comment block. Clicking on that will open a menu with Edit. Then add these lines
```
<!-- BEGIN RELEASE NOTES -->

<!-- END RELEASE NOTES -->
```
and the comment in between them. See the [guide](https://github.com/sympy/sympy/wiki/Writing-Release-Notes) for more details.

**Comment by Saanidhyavats** [comment]
@jksuom and @czgdp1807 thank you for your help. Let me know if there needs to be any other change.

**Comment by Saanidhyavats** [comment]
I have made some additional changes , if any additional changes needs to be done then let me know.

**Comment by oscarbenjamin** [comment]
It looks like the tests have failed on Travis.

**Comment by Saanidhyavats** [comment]
@oscarbenjamin how can I make the travis test successful (whether there is any software which we can install and debug the code accordingly for the test to be successful)

**Comment by oscarbenjamin** [comment]
You should run the tests yourself on your computer and investigate the test failures:
https://github.com/sympy/sympy/wiki/Development-workflow#be-sure-that-all-tests-of-sympy-pass

**Comment by Saanidhyavats** [comment]
Thank you for helping me. I will look into it.

**Comment by Saanidhyavats** [comment]
I ran "./bin/test" on my system for checking the Travis CI test. The test has indicated some issues in other files but not in this one but the same test is showing error here . Whether the procedure followed by me is correct?

**Comment by oscarbenjamin** [comment]
I don't know but the test failures on Travis clearly show an error in your code (you have used `self` in `__new__`). You should try to fix the code so that the tests pass.

**Comment by czgdp1807** [comment]
@Saanidhyavats Please resolve the conflicts. Thanks for your contributions.

**Comment by Saanidhyavats** [comment]
I think this request is ready to be merged @oscarbenjamin

**Comment by oscarbenjamin** [comment]
There should be a regression test for whatever you are changing.

The PR title should be changed to something meaningful. The commits should be squashed and a meaningful commit message should be used.

The release note should describe what the actual effect of the change is for end users rather than what is changed in the code.

**Comment by Saanidhyavats** [comment]
On doing `Derivative(d, x, evaluate = False)` I am getting the expected result as `Derivative(Derivative(y, x), x)` but I don't know why it is failing in the tests. Can someone please help me on this.

File: sympy/core/function.py [hunk 1]
```diff
@@ -1386,10 +1386,10 @@ def __new__(cls, expr, *variables, **kwargs):
             variable_count = cls._sort_variable_count(variable_count)
 
         # denest
-        if isinstance(expr, Derivative):
+        if evaluate is not False and isinstance(expr, Derivative):
```
Comment (czgdp1807) [inline sympy/core/function.py]:
```suggestion
        if evaluate and isinstance(expr, Derivative):
```

File: sympy/core/tests/test_function.py [hunk 2]
```diff
@@ -1355,3 +1355,9 @@ def test_Derivative_free_symbols():
 def test_issue_10503():
     f = exp(x**3)*cos(x**6)
     assert f.series(x, 0, 14) == 1 + x**3 + x**6/2 + x**9/6 - 11*x**12/24 + O(x**14)
+
+
+def test_issue_16350():
+    d = Derivative(y, x)
+    assert Derivative(d, x) == Derivative(y, (x, 2))
+    assert Derivative(d, x, evaluate = False) == Derivative(Derivative(y, x), x)
```
Comment (czgdp1807) [inline sympy/core/tests/test_function.py]:
```suggestion
    assert Derivative(d, x, evaluate=False) == Derivative(Derivative(y, x), x)
```

**Comment by Saanidhyavats** [comment]
@czgdp1807 I am getting the same assert error problem, can you help me on it?

**Comment by czgdp1807** [comment]
```python
>>> d = Derivative(y, x)
>>> Derivative(d, x)
  2   
 d    
───(y)
  2   
dx    
>>> Derivative(d, x) == Derivative(y, (x, 2))
True
>>> Derivative(d, x, evaluate=False) == Derivative(Derivative(y, x), x)
True
```
The issue is fixed in `master`. Closing for now.

**Pull request closed without merging**